File: src/main/java/org/sandboxpowered/silica/state/BaseState.java
package org.sandboxpowered.silica.state;

import com.google.common.collect.ImmutableMap;
import org.sandboxpowered.api.state.property.Property;
import org.sandboxpowered.api.state.property.PropertyContainer;

import java.util.Collection;
import java.util.Iterator;

public class BaseState<A,B> implements PropertyContainer<B> {
    protected final A base;
    private final ImmutableMap<Property<?>, Comparable<?>> properties;

    public BaseState(A base, ImmutableMap<Property<?>, Comparable<?>> properties) {
        this.base = base;
        this.properties = properties;
    }

    @Override
    public <T extends Comparable<T>> T get(Property<T> property) {
        Comparable<?> value = this.properties.get(property);
        if (value == null) {
            throw new IllegalArgumentException(String.format("Cannot get property %s as it does not exist in %s", property, this.base));
        } else {
            return property.getValueType().cast(value);
        }
    }

    @Override
    public <T extends Comparable<T>, V extends T> B with(Property<T> property, V value) {
        Comparable<?> currentValue = this.properties.get(property);
        if (currentValue == null) {
            throw new IllegalArgumentException(String.format("Cannot set property %s as it does not exist in %s", property, this.base));
        } else {
            B state = getState(property, value);
            if (state == null) {
                throw new IllegalArgumentException(String.format("Cannot set property %s to %s on %s, it is not an allowed value", property, value, this.base));
            } else {
                return state;
            }
        }
    }

    @Override
    public <T extends Comparable<T>> B cycle(Property<T> property) {
        Comparable<?> currentValue = this.properties.get(property);
        if (currentValue == null) {
            throw new IllegalArgumentException(String.format("Cannot set property %s as it does not exist in %s", property, this.base));
        }
        return getState(property, findNextInCollection(property.getValues(), (T) currentValue));
    }

    protected static <T> T findNextInCollection(Collection<T> collection, T object) {
        Iterator<T> iterator = collection.iterator();

        do {
            if (!iterator.hasNext()) {
                return iterator.next();
            }
        } while (!iterator.next().equals(object));

        return iterator.hasNext() ? iterator.next() : collection.iterator().next();
    }

    private <T extends Comparable<T>, V extends T> B getState(Property<T> property, V value) {
        return null; // TODO
    }

    @Override
    public <T extends Comparable<T>> boolean contains(Property<T> property) {
        return properties.containsKey(property);
    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/world/state/BaseState.kt
package org.sandboxpowered.silica.world.state

import com.google.common.collect.ArrayTable
import com.google.common.collect.HashBasedTable
import com.google.common.collect.ImmutableMap
import com.google.common.collect.Table
import org.sandboxpowered.silica.registry.RegistryEntry
import org.sandboxpowered.silica.util.extensions.set
import org.sandboxpowered.silica.world.state.property.Property

open class BaseState<B : RegistryEntry<B>, S : PropertyContainer<S>>(
    protected val base: B,
    val properties: ImmutableMap<Property<*>, Comparable<*>>
) : PropertyContainer<S> {

    private lateinit var possibleStates: Table<Property<*>, Comparable<*>, S>

    override fun <T : Comparable<T>> get(property: Property<T>): T {
        val value = properties[property]
        require(value != null) { "Cannot get property $property as it does not exist in $base" }
        return property.valueType.cast(value)
    }

    override fun <T : Comparable<T>> set(property: Property<T>, value: T): S {
        val currentValue = properties[property]
        require(currentValue != null) { "Cannot set property $property as it does not exist in $base" }
        if (currentValue == value) return this as S
        val state = getState(property, value)
        return state ?: error("Cannot set property $property to $value on $base, it is not an allowed value")
    }

    override fun <T : Comparable<T>> contains(property: Property<T>): Boolean = property in properties

    override fun <T : Comparable<T>> cycle(property: Property<T>): S {
        val currentValue = properties[property] ?: error("Cannot set property $property as it does not exist in $base")
        return set(property, findNextInCollection(property.values, currentValue as T))
    }

    fun initTable(map: Map<Map<Property<*>, Comparable<*>>, S>) {
        val table: Table<Property<*>, Comparable<*>, S> = HashBasedTable.create()
        for ((property, value) in properties) {
            val comparableIterator = property.values.iterator()
            while (comparableIterator.hasNext()) {
                val comparable = comparableIterator.next()
                if (comparable != value) {
                    map[createPropertiesCollectionWith(property, comparable)]?.let {
                        table[property, comparable] = it
                    }
                }
            }
        }
        possibleStates = if (table.isEmpty) table else ArrayTable.create(table)
    }

    private fun createPropertiesCollectionWith(property: Property<*>, comparable: Comparable<*>) =
        HashMap(properties).apply { set(property, comparable) }

    private fun <T : Comparable<T>, V : T> getState(property: Property<T>, value: V): S? =
        possibleStates[property, value] // TODO: replace with global state holder object rather than holding every possible state within all states.

    override fun toString(): String = "${javaClass.simpleName}{base=$base, properties=$properties}"

    private fun <T> findNextInCollection(collection: Collection<T>, value: T): T {
        val iterator = collection.iterator()
        do {
            if (!iterator.hasNext()) {
                return iterator.next()
            }
        } while (iterator.next() != value)
        return if (iterator.hasNext()) iterator.next() else collection.iterator().next()
    }

    fun modify(block: StateModifier<S>.() -> Unit): S {
        return StateModifier(this).apply(block).state as S
    }
}

class StateModifier<V : PropertyContainer<V>>(var state: PropertyContainer<V>) {
    infix fun <T : Comparable<T>> Property<T>.set(value: T) {
        state = state.set(this, value)
    }
}